### PR TITLE
Performance improvement in cudf::strings::join_strings for long strings

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Improves performance for longer strings with `cudf::strings::join_strings()` API.

The new implementation is inspired by recent changes in json/writer_impl.cu which uses the `make_strings_children` factor that accepts an iterator of string pointers to gather the results into a single column of data but does lose some performance for smallish-strings (less than 32 bytes). So the original code is refactored to work on smaller strings.
Also, a benchmark implementation is added in this PR to test against various sized input columns.
For longer strings, the performance improvement is up to 100x for very long strings (>512 bytes).

Reference https://github.com/rapidsai/cudf/issues/13048

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/string/join_strings.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <benchmarks/common/generate_input.hpp>
+#include <benchmarks/fixture/rmm_pool_raii.hpp>
```
Comment (PointKernel) [inline cpp/benchmarks/string/join_strings.cpp:18]:
```suggestion
```

File: cpp/benchmarks/string/join_strings.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+  data_profile const table_profile = data_profile_builder().distribution(
+    cudf::type_id::STRING, distribution_id::NORMAL, 0, row_width);
+  auto const table =
+    create_random_table({cudf::type_id::STRING}, row_count{num_rows}, table_profile);
+  cudf::strings_column_view input(table->view().column(0));
+
+  state.set_cuda_stream(nvbench::make_cuda_stream_view(cudf::get_default_stream().value()));
+  // gather some throughput statistics as well
+  auto chars_size = input.chars_size();
```
Comment (PointKernel) [inline cpp/benchmarks/string/join_strings.cpp:44]:
```suggestion
  auto const chars_size = input.chars_size();
```

File: cpp/src/strings/combine/join.cu [hunk 3]
```diff
@@ -33,95 +33,142 @@
...
+ * This value was found using the strings_join benchmark results.
+ */
+constexpr size_type AVG_CHAR_BYTES_THRESHOLD = 32;
+
+struct join_base_fn {
+  column_device_view const d_strings;
+  string_view d_separator;
+  string_scalar_device_view d_narep;
+
+  __device__ std::pair<string_view, string_view> process_string(size_type idx) const
```
Comment (PointKernel) [inline cpp/src/strings/combine/join.cu:61]: maybe `thrust::pair` since it's a device code

**Review by PointKernel — APPROVED** [review]
Some non-blocking nits otherwise LGTM

File: cpp/src/strings/combine/join.cu [hunk 4]
```diff
@@ -33,95 +33,142 @@
...
+ * Otherwise, a regular string-parallel function is used.
+ *
+ * This value was found using the strings_join benchmark results.
+ */
+constexpr size_type AVG_CHAR_BYTES_THRESHOLD = 32;
+
+struct join_base_fn {
+  column_device_view const d_strings;
+  string_view d_separator;
+  string_scalar_device_view d_narep;
```
Comment (ttnghia) [inline cpp/src/strings/combine/join.cu:59]:
```suggestion
  string_view const d_separator;
  string_scalar_device_view const d_narep;
```

File: cpp/src/strings/combine/join.cu [hunk 5]
```diff
@@ -33,95 +33,142 @@
...
+  join_fn(column_device_view const d_strings,
+          string_view d_separator,
+          string_scalar_device_view d_narep)
+    : join_base_fn{d_strings, d_separator, d_narep}
+  {
+  }
+
+  __device__ void operator()(size_type idx) const
+  {
+    auto [d_str, d_sep] = process_string(idx);
```
Comment (ttnghia) [inline cpp/src/strings/combine/join.cu:97]:
```suggestion
    auto cosnt [d_str, d_sep] = process_string(idx);
```

File: cpp/src/strings/combine/join.cu [hunk 6]
```diff
@@ -33,95 +33,142 @@
...
+  join_gather_fn(column_device_view const d_strings,
+                 string_view d_separator,
+                 string_scalar_device_view d_narep)
+    : join_base_fn{d_strings, d_separator, d_narep}
+  {
+  }
+
+  __device__ string_index_pair operator()(size_type idx) const
+  {
+    auto [d_str, d_sep] = process_string(idx / 2);
```
Comment (ttnghia) [inline cpp/src/strings/combine/join.cu:121]:
```suggestion
    auto const [d_str, d_sep] = process_string(idx / 2);
```

File: cpp/src/strings/combine/join.cu [hunk 7]
```diff
@@ -33,95 +33,142 @@
...
+    auto indices = cudf::detail::make_counting_transform_iterator(
+      0, join_gather_fn{*d_strings, d_separator, d_narep});
+    auto joined_col = make_strings_column(indices, indices + (input.size() * 2), stream, mr);
+    return std::move(joined_col->release().children.back());
+  }();
+
+  // build the offsets: single string output has offsets [0,chars-size]
+  rmm::device_uvector<size_type> offsets(2, stream);
+  offsets.set_element_to_zero_async(0, stream);
+  offsets.set_element(1, chars_column->size(), stream);
```
Comment (ttnghia) [inline cpp/src/strings/combine/join.cu:161]:
```suggestion
  auto offsets = make_device_uvector_async(std::vector<size_type>{0, chars_column->size()}, stream);
```

File: cpp/src/strings/combine/join.cu (hunk 4 above)
Comment (davidwendt) [reply cpp/src/strings/combine/join.cu:59]: I think adding `const` to `view` member variables is redundant and unnecessary.

File: cpp/src/strings/combine/join.cu [hunk 8]
```diff
@@ -156,9 +157,10 @@ std::unique_ptr<column> join_strings(strings_column_view const& input,
   }();
 
   // build the offsets: single string output has offsets [0,chars-size]
-  rmm::device_uvector<size_type> offsets(2, stream);
-  offsets.set_element_to_zero_async(0, stream);
-  offsets.set_element(1, chars_column->size(), stream);
+  auto offsets =
+    cudf::detail::make_device_uvector_async(std::vector<size_type>({0, chars_column->size()}),
+                                            stream,
+                                            rmm::mr::get_current_device_resource());
```
Comment (ttnghia) [inline cpp/src/strings/combine/join.cu:163]: `mr`?

**Pull request merged**